Mathlib/Data/List/TakeDrop.lean

```diff
@@ -151,6 +151,71 @@ theorem span_eq_takeWhile_dropWhile (l : List α) : span p l = (takeWhile p l, d
 
 end Filter
 
+section Splits
+
+/-! ### Generate splits of a list. -/
+
+/-- Pairs of prefixes and suffixes of a list. -/
+def splits (xs : List α) : List (List α × List α) :=
+  List.map xs.splitAt (range (xs.length + 1))
+
+@[simp]
+lemma splits_nil : (List.nil (α:=α)).splits= [([], [])] := by rfl
+
+@[simp]
+lemma splits_cons (x : α) (xs : List α) :
+    List.splits (x :: xs) =
+    ([], x :: xs) :: List.map (fun td ↦ (x :: td.1, td.2)) (List.splits xs) := by
+  simp [List.splits, List.range_succ_eq_map]
+
+lemma mem_splits_iff (x y l : List α) :
+    (x, y) ∈ l.splits ↔ l = x ++ y := by
+  simp only [splits, splitAt_eq, mem_map, mem_range, Prod.mk.injEq]
+  constructor
+  · rintro ⟨n, hn, htake, hdrop⟩
+    rw [←htake, ←hdrop, List.take_append_drop]
+  · rintro rfl
+    exists x.length
+    constructor
+    · simp only [length_append]; omega
+    · constructor
+      · rw [List.take_left]
+      · rw [List.drop_left]
+
+lemma nil_self_mem_splits (l : List α) : ([], l) ∈ l.splits := by
+  simp [List.mem_splits_iff]
+
+lemma self_nil_mem_splits (l : List α) : (l, []) ∈ l.splits := by
+  simp [List.mem_splits_iff]
+
+/-- Left-associative triple splits of a list. -/
+def splits₃_left : List α → List (List α × List α × List α) :=
+  List.flatMap
+  (fun td ↦
+    td.1
+    |> List.splits
+    |> List.map (fun t ↦ (t.1, t.2, td.2)))
+  ∘ List.splits
```

Comment:
```suggestion
/-- Left-associative triple splits of a list. -/
def splits₃Left (l : List α) : List (List α × List α × List α) :=
  l.splits.flatMap fun td ↦ td.1.splits.map fun t ↦ (t.1, t.2, td.2)
```
In the naming convention, definitions cannot contain underscores. I've also used dot notation to clean up the code. We also prefer to include everything to the left of the colon if possible, instead of using composition.